rasterio/rio/features.py

```diff
@@ -118,6 +119,20 @@ def mask(
         bounds = geojson.get('bbox', calculate_bounds(geojson))
 
         with rasterio.open(input) as src:
+            
+            # Determine the projection of the GeoJSON mask
+            if geojson.get('crs', {}).get('type') == 'name':
+                from_crs = geojson.get('crs').get('properties').get('name')
+            else:
+                from_crs = {'init': 'epsg:4326'}
+            
+            # Transform input bounds and geometries to the src projection
+            x, y = bounds[0::2], bounds[1::2]
+            (left, right), (bottom, top) = rasterio.warp.transform(from_crs, src.crs, x, y)
+            bounds = BoundingBox(left, bottom, right, top)
```

Comment:
You should get new bounds after transforming the geometry, not project the bounds you get from the geometry before projecting them.

I think you could move this whole block outside the ```with``` block, and get the bounds exactly once after transforming the geojson if needed.